ntoskrnl/cc/cacheman.c

```diff
@@ -137,17 +138,20 @@ CcGetFlushedValidData (
 }
 
 /*
- * @unimplemented
+ * @implemented
  */
 PVOID
 NTAPI
 CcRemapBcb (
     IN PVOID Bcb
     )
 {
-	UNIMPLEMENTED;
+    PINTERNAL_BCB iBcb = Bcb;
+
+    CCTRACE(CC_API_DEBUG, "Bcb=%p\n", Bcb);
 
-    return 0;
+    iBcb->RefCount++;
```

Comment:
Every `RefCount` increase must have a corresponding `RefCount` decrease.
Where is the decrease for this one?

Looks like this stub can just be improved by returning `Bcb` instead of `NULL`

Reply:
It probably should be done on failure case, which I didn't implement yet.

Reply:
No @Extravert-ir , just returning `Bcb` instead of `0` does not work for me. The system fails again for me when trying to write anything to NTFS partition. :frowning_face: The only difference is some asserts, whose don't appear if to return `0`.
And moreover, if to look at `CcRepinBcb` implementation (which is similar to `CcRemapBcb`), it also has only increase, but does not have any decrease. And it's not used by anything other (in the kernel) that has appropriate decrease, same as my one. Either this routine indeed shouldn't be used by any other ones, or those ones are just not implemented yet. But in any case, if it works at least for NTFS driver, then my implementation is definitely correct.
So your argument seems to be wrong in this case.

Reply:
The MSDN documentation is unclear about it : should a call to CcRemapBcb be paired with CcUnpinData ?

Reply:
I found the decrease one for `CcRepinBcb`. It is called many times it some our FS drivers, e. g. fastfat_new, for example here: https://git.reactos.org/?p=reactos.git;a=blob;f=drivers/filesystems/fastfat_new/allocsup.c;hb=8ee88d3bd5bb2464424490bcbca8510d2116df51#l4117, and each time, after that, it calls `CcUnpinData`, whose implementation returns `CcUnpinDataForThread`, which, in its turn, contains a decrease for reference count of `Bcb`: https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/cc/pin.c;hb=8ee88d3bd5bb2464424490bcbca8510d2116df51#l617.
It's probably done by the same way in MS NTFS driver too. First it calls `CcRemapBcb` somewhere, and then some appropriate routine for it, which decrements the reference count (probably `CcUnpinData` or other). :slightly_smiling_face: 
So looks like the decrease is done by another function (currently implemented or not), and it's not needed to add it into my implementation.
See https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/cc/pin.c;hb=8ee88d3bd5bb2464424490bcbca8510d2116df51#l155 also (called by `CcUnpinData`).

Reply:
> The MSDN documentation is unclear about it : should a call to CcRemapBcb be paired with CcUnpinData ?

Probably not, but there definitely should be some routine for decrease. Judging by MSDN documentation, and our code, it's `CcUnpinRepinnedBcb`: https://git.reactos.org/?p=reactos.git;a=blob;f=ntoskrnl/cc/pin.c;hb=8ee88d3bd5bb2464424490bcbca8510d2116df51#l642.